lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,369 @@
+
+import iris
+import numpy as np
+from numpy.linalg import lstsq
+from improver.utilities.cube_manipulation import enforce_float32_precision
+from improver.constants import DALR
+
+from scipy.ndimage import generic_filter
+
+
+class SaveNeighbourhood(object):
+    """Saves the neighbourhood around each central point.
+
+    The "generic_filter" module extracts the neighbourhood around each
+    point as "buffer". This buffer is passed to the "SaveNeighbourhood" class.
+    The "filter" function then saves this buffer into the "allbuffers" array.
+
+    """
+    def __init__(self, allbuffers):
+        """Initialise the class.
+
+        Create the global variables that allows the "filter" function
+        to save each extracted buffer into "allbuffers".
+
+        """
+        # Initialises the iterator.
+        self.i = 0
+        # Saves all the buffers into this array.
+        self.allbuffers = allbuffers
+
+    def filter(self, buffer):
+        """Defines filter function to be applied to extracted buffers.
+
+        Saves the contents of the buffer into "allbuffers" array. Therefore
+        a return value isn't required. However "generic_filter"
+        requires a return value - so use zero.
+
+        Args:
+            buffer (array):
+                Array containing neighourbood points.
+
+        Returns:
+            zero (float)
+                Blank return value required by "generic_filter".
+
+        """
+        self.allbuffers[self.i, :] = buffer
+        self.i += 1
+        return 0.0
+
+
+class LapseRate(object):
+    """Plugin to calculate the lapse rate from orography and temperature
+       cubes.
+
+    References:
+        The method applied here is based on the method used in the 2010 paper:
+        https://rmets.onlinelibrary.wiley.com/doi/abs/10.1002/met.177
+
+    Code methodology:
+    1) Apply land/sea mask to temperature and orography datasets. Mask sea
+       points as NaN since image processing module does not recognise Numpy
+       masks.
+    2) Extracts neighbourhoods from both datasets:
+       Apply "generic_filter" image processing module to temperature data
+       to extract a neighbourhood for each single point and save this
+       neighbourhood into a larger array.
+       Repeat for orography data.
+       generic_filter mode="constant" ensures that points beyond edges of
+       dataset will be filled with "cval". (NaN in this case)
+    3) For all the stored orography neighbourhoods - take the neighbours around
+       the central point and create a mask where the height difference from
+       the central point is greater than 35m.
+    4) Loop through array of neighbourhoods and take the height and temperature
+       of all grid points and calculate the
+       temperature/height gradient = lapse rate
+    5) Constrain the lapse rate as > DALR and < -3.0*DALR
+
+    """
+
+    def __init__(self, max_height_diff=35, nbhood_radius=1,
```

Comment:
No real basis. The code in the research paper was a 4x4 neighbourhood for a 4km model. So I thought it was best to start with the minimum radius=1 to give a 3x3 grid. We'll probably want to change this later. It would be good to run trials to see what the best radius is -  I'll put a comment in the JIRA ticket.

Reply:
I've put in a default nbhood_radius=3 into the plugin. The test code will still use nbhood_radius=1 so that we don't have to have larger arrays in the test code (which would make it much less readable).

Reply:
I think the paper was 8x8 grid squares for a 4 km model, which makes the equivalent radius 7?  Could we use `nbhood_radius=7` as the default please?  And could you add a quick note to the docstring on where this default has come from (and the 35 km height - just say it's come from the paper referenced).

Thanks.

Reply:
Done - the good news is that the larger neighbourhoods don't slow down the code significantly (a few seconds). This seems to point more to the loop itself being the bottleneck rather than calculating the data within it.